docker/util/run.bzl

```
@@ -277,6 +285,186 @@ commit = struct(
     implementation = _commit_impl,
 )
 
+def _commit_layer_impl(
+        ctx,
+        name = None,
+        image = None,
+        commands = None,
+        docker_run_flags = None,
+        env = None,
+        compression = None,
+        compression_options = None,
+        output_layer_tar = None):
+    """Implementation for the container_run_and_commit_layer rule.
+
+    This rule runs a set of commands in a given image, waits for the commands
+    to finish, and then commits the container to a new image.
+
+    Args:
+        ctx: The bazel rule context
+        name: A unique name for this rule.
+        image: The input image tarball
+        commands: The commands to run in the input image container
+        docker_run_flags: String list, overrides ctx.attr.docker_run_flags
+        env: str Dict, overrides ctx.attr.env
+        compression: str, overrides ctx.attr.compression
+        compression_options: str list, overrides ctx.attr.compression_options
+        output_layer_tar: The output layer obtained as a result of running
+                          the commands on the input image
+    """
+
+    name = name or ctx.attr.name
+    image = image or ctx.file.image
+    commands = commands or ctx.attr.commands
+    docker_run_flags = docker_run_flags or ctx.attr.docker_run_flags
+    env = env or ctx.attr.env
+    script = ctx.actions.declare_file(name + ".build")
+    compression = compression or ctx.attr.compression
+    compression_options = compression_options or ctx.attr.compression_options
+    output_layer_tar = output_layer_tar or ctx.outputs.layer
+
+    toolchain_info = ctx.toolchains["@io_bazel_rules_docker//toolchains/docker:toolchain_type"].info
+
+    # Generate a shell script to execute the reset cmd
+    image_utils = ctx.actions.declare_file("image_util.sh")
+    ctx.actions.expand_template(
+        template = ctx.file._image_utils_tpl,
+        output = image_utils,
+        substitutions = {
+            "%{docker_flags}": " ".join(toolchain_info.docker_flags),
+            "%{docker_tool_path}": toolchain_info.tool_path,
+        },
+        is_executable = True,
+    )
+
+    docker_env = [
+        "{}={}".format(
+            ctx.expand_make_variables("env", key, {}),
+            ctx.expand_make_variables("env", value, {}),
+        )
+        for key, value in env
```

When I specify the `env` argument, I get `type 'string' is not iterable`. This line should probably be `for key, value in env.items()`

---

oh interesting, I wonder how it worked for me 🤔 Thanks!
```suggestion
        for key, value in env.items()
```